src/kill-parallel.c

```diff
@@ -0,0 +1,261 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <errno.h>
+#include <pthread.h>
+#include <unistd.h>
+#include <string.h>
+#include <signal.h>
+#include <poll.h>
+#include <Rinternals.h>
+#include "posix.h"
+#include "ps-internal.h"
+
+// Should that be a global option shared by processx and ps?
+#define PS_INTERRUPT_INTERVAL 200
+
+static struct sigaction old_sig_handler;
+static int kill_parallel_pipe = -1;
+
+static void kill_parallel_sigchld_callback(int sig, siginfo_t *info, void *ctx);
+static void kill_parallel_setup_sigchld(struct sigaction *prev);
+static void kill_parallel_remove_sigchld(struct sigaction *prev);
+static void block_sigchld_save(sigset_t *old);
+static void procmask_set(sigset_t *set);
+static void unblock_sigchld(void);
+
+SEXP ps__kill_parallel(SEXP ps, SEXP ffi_grace) {
+  SEXP cont = PROTECT(R_MakeUnwindCont());
+  SEXP err = NULL;
+
+  sigset_t old_mask;
+  struct sigaction old_action;
+
+  block_sigchld_save(&old_mask);
+  kill_parallel_setup_sigchld(&old_action);
+
+  int ps_n = length(ps);
+
+  SEXP killed_shelter = PROTECT(allocVector(LGLSXP, ps_n));
+  int* killed = LOGICAL(killed_shelter);
+
+  int killed_n = 0;
+  memset(killed, 0, sizeof(int) * ps_n);
+
+  int fds[2] = { -1 };
+  if (pipe(fds)) {
+    ps__set_error("Can't create pipe for parallel kill.");
+    err = ps__last_error;
+    goto cleanup;
+  }
+  kill_parallel_pipe = fds[1];
+
+  ps__nonblock_fcntl(fds[0], 1);
+  ps__nonblock_fcntl(fds[1], 1);
+
+  for (int i = 0; i < ps_n; ++i) {
+    ps_handle_t *handle = R_ExternalPtrAddr(VECTOR_ELT(ps, i));
+    if (psll__is_running(handle)) {
+      kill(handle->pid, SIGTERM);
```

Comment:
For the record, this is a race condition here, the comments from `api-posix.c` apply here as well:
```
  /* TODO: this is still a race here. We would need to SIGSTOP the
     process first, then check the timestamp, and then send the signal
     (if not SIGSTOP), or send a SIGCONT. */
```